funsor/terms.py

```diff
@@ -171,6 +188,60 @@ def __call__(cls, *args, **kwargs):
 
         return interpret(cls, *args)
 
+    def __getitem__(cls, arg_types):
+        if not isinstance(arg_types, tuple):
+            arg_types = (arg_types,)
+        assert not any(isvariadic(arg_type) for arg_type in arg_types), "nested variadic types not supported"
+        # TODO add support for nested union types (i.e. remove the following assertion)
+        assert not any(isinstance(arg_type, tuple) for arg_type in arg_types), "nested union types not supported"
+        # switch tuple to typing.Tuple
+        arg_types = tuple(typing.Tuple if arg_type is tuple else arg_type for arg_type in arg_types)
+        if arg_types not in cls._type_cache:
+            assert not cls.__args__, "cannot subscript a subscripted type {}".format(cls)
+            assert len(arg_types) == len(cls._ast_fields), "must provide types for all params"
+            new_name = cls.__name__ + "[{}]".format(", ".join(
+                t.__name__ if hasattr(t, "__name__") else str(t) for t in arg_types))  # Tuple doesn't have __name__
+            new_dct = cls.__dict__.copy()
+            new_dct.update({"__args__": arg_types})
+            # type(cls) to handle FunsorMeta subclasses
+            cls._type_cache[arg_types] = type(cls)(new_name, (cls,), new_dct)
+        return cls._type_cache[arg_types]
+
+    def __subclasscheck__(cls, subcls):  # issubclass(subcls, cls)
+        if not isinstance(subcls, FunsorMeta):
+            return super(FunsorMeta, getattr(cls, "__origin__", cls)).__subclasscheck__(subcls)
+        elif cls.__args__ and subcls.__args__:
+            return subcls is cls or \
+                (super(FunsorMeta, cls.__origin__).__subclasscheck__(subcls.__origin__) and
+                 len(cls.__args__) == len(subcls.__args__) and
+                 all(_issubclass_tuple(subcls_param, param)
+                     for subcls_param, param in zip(subcls.__args__, cls.__args__)))
+        elif not cls.__args__ and subcls.__args__:
+            return super(FunsorMeta, cls).__subclasscheck__(subcls.__origin__)
+        elif cls.__args__ and not subcls.__args__:
+            # issubclass(Number, Number[int, int]) == False
+            # issubclass(SubclassOfNumber, Number[int, int]) == True
+            return cls.__origin__ is not subcls and super(FunsorMeta, cls.__origin__).__subclasscheck__(subcls)
+        elif not cls.__args__ and not subcls.__args__:
+            return super(FunsorMeta, cls).__subclasscheck__(subcls)
+        raise TypeError("could not unify: {}, {}".format(cls, subcls))
+
+
+def _issubclass_tuple(subcls, cls):
+    """
+    utility for pattern matching with tuple subexpressions
+    """
+    subcls_is_tuple = hasattr(cls, "__origin__") and subcls.__origin__ is tuple
+    cls_is_tuple = hasattr(cls, "__origin__") and cls.__origin__ is tuple
+    if not subcls_is_tuple and not cls_is_tuple:
```

Comment:
nit: again, I find it a bit easier to read the early-exit pattern. maybe something like
```py
if subclas_is_tuple != cls_is_tuple:
    return False
if not cls_is_tuple:
    return issubclass(subcls, cls)
if not cls.__args__:
    return True
if len(subcls.__args__) != len(cls.__args__)
    return False
for a, b in zip(subcls.__args__, cls.__args__):
    if not _issubclass_tuple(a, b):
        return False
return True
```